### PR TITLE
[Merged by Bors] - feat(AlgebraicGeometry/EllipticCurve/J): further properties of j-invariants of elliptic curves

### PR DESCRIPTION
**Pull request by acmepjz**
Currently there is only one new result:

- `EllipticCurve.exists_variableChange_of_j_eq`: if `E` and `E'` are elliptic curves with the same `j`-invariants defined over a separably closed field, then there exists a change of variables over that field which change `E` into `E'`.

For convenience, a few new results for normal forms are added to `EllipticCurve` namespace, parallel to that of `WeierstrassCurve`.

Also added some convenience auxiliary results used in this PR:

- `Algebra/CharP/Defs`: `CharP.cast_ne_zero_of_ne_of_prime`
- `FieldTheory/Separable`: `Polynomial.separable_C_mul_X_pow_add_C_mul_X_add_C[']`
- `FieldTheory/IsSepClosed`: `IsSepClosed.exists_root_C_mul_X_pow_add_C_mul_X_add_C[']`

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

discussion: https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/.2317431.20further.20properties.20of.20j-invariants/near/474955205

### CODE REVIEW

File: Mathlib/AlgebraicGeometry/EllipticCurve/J.lean [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+
+- `EllipticCurve.exists_variableChange_of_j_eq`: if `E` and `E'` are elliptic curves with the same
+  `j`-invariants defined over a separably closed field, then there exists a change of variables
+  over that field which change `E` into `E'`.
+
+-/
+
+open WeierstrassCurve Polynomial
+
+-- TODO: move to suitable location ???
```
Comment (Ruben-VandeVelde) [inline Mathlib/AlgebraicGeometry/EllipticCurve/J.lean:25]: Sounds like a good idea, maybe Mathlib.FieldTheory.Separable

File: Mathlib/AlgebraicGeometry/EllipticCurve/J.lean [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+    simp_rw [f, map_add derivative, derivative_C]
+    simp [show (n : R) = 0 from (CharP.cast_eq_zero_iff R p n).2 hdvd]
+  obtain ⟨e, ha⟩ := ha.exists_left_inv
+  refine ⟨-derivative f, f + C e, ?_⟩
+  rw [hderiv, right_distrib, ← add_assoc, neg_mul, mul_comm, neg_add_cancel, zero_add,
+    ← map_mul, ha, map_one]
+
+variable {F : Type*} [Field F] [IsSepClosed F]
+
+-- TODO: move to suitable location ???
```
Comment (Ruben-VandeVelde) [inline Mathlib/AlgebraicGeometry/EllipticCurve/J.lean:41]: Mathlib.FieldTheory.IsSepClosed

File: Mathlib/AlgebraicGeometry/EllipticCurve/J.lean [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+    have h1 : n ≠ 1 := by linarith only [hn]
+    have : 1 ≤ n := le_trans one_le_two hn
+    compute_degree!
+    simp [h0, h1]
+  have hsep : f.Separable := separable_X_pow_add_C_mul_X_add_C p n hdvd a b ha.isUnit
+  obtain ⟨x, hx⟩ := IsSepClosed.exists_root f hdeg hsep
+  exact ⟨x, by simpa [f] using hx⟩
+
+-- TODO: move to suitable location ???
+private lemma CharP.cast_ne_zero_of_ne_of_prime {R : Type*} [NonAssocSemiring R]
```
Comment (Ruben-VandeVelde) [inline Mathlib/AlgebraicGeometry/EllipticCurve/J.lean:60]: Mathlib.Algebra.CharP.Defs

File: Mathlib/AlgebraicGeometry/EllipticCurve/J.lean (hunk 1 above)
Comment (acmepjz) [reply Mathlib/AlgebraicGeometry/EllipticCurve/J.lean]: Done.

File: Mathlib/AlgebraicGeometry/EllipticCurve/J.lean (hunk 2 above)
Comment (acmepjz) [reply Mathlib/AlgebraicGeometry/EllipticCurve/J.lean]: Done.

File: Mathlib/AlgebraicGeometry/EllipticCurve/J.lean (hunk 3 above)
Comment (acmepjz) [reply Mathlib/AlgebraicGeometry/EllipticCurve/J.lean]: Done.

**Review by jcommelin — COMMENTED** [review]
I am going to kick this on the merge queue now, because this is a very nice result.

I do wonder whether some of the constructions hidden inside the proofs of these private lemmas should at some point be turned into definitions.

Will there be cases in the future where someone has two elliptic curves over a field that is not char 2 or char 3, and they want to reason hands on with an explicit change of variables?

At the moment they only get the mere existence of such.

bors merge

**Comment by acmepjz** [comment]
> I do wonder whether some of the constructions hidden inside the proofs of these private lemmas should at some point be turned into definitions.
> 
> Will there be cases in the future where someone has two elliptic curves over a field that is not char 2 or char 3, and they want to reason hands on with an explicit change of variables?
> 
> At the moment they only get the mere existence of such.

Yes that's possible.

But according to the proof listed in Silverman's GTM106, Chapter III, Prop. 1.4(b), that definition will depends on the j-invariant of the elliptic curve (being 0, 1728 or otherwise), and taking square root, 4th root or 6th root. I'm not sure if it should be explicitly written (it needs Classical.choice for choosing a root in separably closed field anyway...)

Anyway, if someone founds that the explicit definition is useful, it can always be added back later.

**Pull request closed without merging**